pkg/boot/linux_test.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2017-2020 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package boot
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/u-root/u-root/pkg/curl"
+	"github.com/u-root/u-root/pkg/uio"
+)
+
+func TestLinuxLabel(t *testing.T) {
+	dir, err := ioutil.TempDir("", "foo")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	osKernel, err := os.Create(filepath.Join(dir, "kernel"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	osInitrd, err := os.Create(filepath.Join(dir, "initrd"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	k, _ := url.Parse("http://127.0.0.1/kernel")
+	i1, _ := url.Parse("http://127.0.0.1/initrd1")
+	i2, _ := url.Parse("http://127.0.0.1/initrd2")
+	httpKernel, _ := curl.LazyFetch(k)
+	httpInitrd1, _ := curl.LazyFetch(i1)
+	httpInitrd2, _ := curl.LazyFetch(i2)
+
+	for _, tt := range []struct {
+		desc string
+		img  *LinuxImage
+		want string
+	}{
+		{
+			desc: "os.File",
+			img: &LinuxImage{
+				Kernel: osKernel,
+				Initrd: osInitrd,
+			},
+			want: fmt.Sprintf("Linux(kernel=%s/kernel, initrd=%s/initrd)", dir, dir),
+		},
+		{
+			desc: "lazy file",
+			img: &LinuxImage{
+				Kernel: uio.NewLazyFile(filepath.Join(dir, "kernel")),
+				Initrd: uio.NewLazyFile(filepath.Join(dir, "initrd")),
+			},
+			want: fmt.Sprintf("Linux(kernel=%s/kernel, initrd=%s/initrd)", dir, dir),
+		},
+		{
+			desc: "concat lazy file",
+			img: &LinuxImage{
+				Kernel: uio.NewLazyFile(filepath.Join(dir, "kernel")),
+				Initrd: CatInitrds(
+					uio.NewLazyFile(filepath.Join(dir, "initrd")),
+					uio.NewLazyFile(filepath.Join(dir, "initrd")),
+				),
+			},
+			want: fmt.Sprintf("Linux(kernel=%s/kernel, initrd=%s/initrd, %s/initrd)", dir, dir, dir),
+		},
+		{
+			desc: "curl lazy file",
+			img: &LinuxImage{
+				Kernel: httpKernel,
+				Initrd: CatInitrds(
+					httpInitrd1,
+					httpInitrd2,
+				),
+			},
+			want: "Linux(kernel=http://127.0.0.1/kernel, initrd=http://127.0.0.1/initrd1, http://127.0.0.1/initrd2)",
```

Comment:
Looking at this specific example, I would suggest to either prefix every initrd with `initrd=`, use a different separator than `, ` for the initrds, possibly just removing the space so it's clearer that they belong together, or a different separator between kernel and initrds, e.g., a semicolon. The first and second option would be more consistent with other notations like the kernel command-line and syslinux' (dropping the space, like in https://wiki.syslinux.org/wiki/index.php?title=Config#INITRD), respectively. Maybe I am overthinking it though.

Reply:
good point, i did both